lib/quicly.c

```diff
@@ -4333,9 +4350,20 @@ static int do_send(quicly_conn_t *conn, quicly_send_context_t *s)
     /* handle timeouts */
     if (conn->idle_timeout.at <= conn->stash.now) {
         QUICLY_PROBE(IDLE_TIMEOUT, conn, conn->stash.now);
-        conn->super.state = QUICLY_STATE_DRAINING;
-        destroy_all_streams(conn, 0, 0);
-        return QUICLY_ERROR_FREE_CONNECTION;
+        goto CloseNow;
+    }
+    /* handle handshake timeouts */
+    if ((conn->initial != NULL || conn->handshake != NULL) &&
+        conn->created_at + (uint64_t)conn->super.ctx->handshake_timeout_rtt_multiplier * conn->egress.loss.rtt.smoothed <=
+            conn->stash.now) {
+        QUICLY_PROBE(HANDSHAKE_TIMEOUT, conn, conn->stash.now, conn->stash.now - conn->created_at, conn->egress.loss.rtt.smoothed);
+        conn->super.stats.num_handshake_timeouts++;
+        goto CloseNow;
+    }
+    if (conn->super.stats.num_packets.initial_handshake_sent >= conn->super.ctx->max_initial_handshake_packets) {
```

Comment:
```suggestion
    if (conn->super.stats.num_packets.initial_handshake_sent > conn->super.ctx->max_initial_handshake_packets) {
```
I think other "max" values retained by `quicly_context_t` are inclusive?